source/extensions/filters/listener/original_src/original_src_socket_option.cc

```diff
@@ -0,0 +1,79 @@
+#include "extensions/filters/listener/original_src/original_src_socket_option.h"
+
+#include <arpa/inet.h>
+
+#include "common/common/assert.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace ListenerFilters {
+namespace OriginalSrc {
+
+constexpr uint8_t OriginalSrcSocketOption::IPV4_KEY;
+constexpr uint8_t OriginalSrcSocketOption::IPV6_KEY;
+
+OriginalSrcSocketOption::OriginalSrcSocketOption(
+    Network::Address::InstanceConstSharedPtr src_address)
+    : src_address_(std::move(src_address)) {
+  // Source transparency only works on IP connections.
+  ASSERT(src_address_->type() == Network::Address::Type::Ip);
+}
+
+bool OriginalSrcSocketOption::setOption(
+    Network::Socket& socket, envoy::api::v2::core::SocketOption::SocketState state) const {
+
+  if (state == envoy::api::v2::core::SocketOption::STATE_PREBIND) {
+    socket.setLocalAddress(src_address_);
+  }
+
+  bool result = true;
+  std::for_each(options_to_apply_.begin(), options_to_apply_.end(),
```

Comment:
just use range for? what's the benefit of for_each here?

Reply:
I prefer range-for slightly here for the simplicity and easy to read. for_each is for complicated case with different than default begin/end.

Reply:
I'll change it.